File: src/SumIO.cpp
// IO functions of cpp and inimate using haskell code
// Min Zhang
// Feb 11, 2016

#include <iostream>

using namespace std;

int main(){
  int age;
  int numberOfPeople = 0;
  int sumOfAge = 0;

  cout << "Input age, type -1 if you want to end the program." << endl;
  cin >> age;

  while (age != -1){
    sumOfAge += age;
    numberOfPeople ++;
    cout << "type another age, -1 to end: " << endl;
    cin >> age; 
  }

  cout << "total: " << numberOfPeople << " and average age is: " << sumOfAge/numberOfPeople << endl;
  
  return 0;
}

File: src/SumIO.hs
-- Feb 11, 2016
-- same IO program as cpp code, in Haskell
-- it seems like much less efficient than cpp code
-- compiled cpp exe is only 16k, compared to haskell exe 1.8MB
-- but the good thing is: all the data is preserved, you can do more computation if necessary, but cpp code is mutated and raw data is lost.


-- both codes are easily broken if the input age is not a number

-- So in long term, Haskell code is definitely safer and might be more flexible and efficient.

import Control.Applicative
import qualified Safe 

main = do
  allages <- getAge
  putStrLn (show $ sum allages)
  putStrLn (show $ getAverage allages)

getAge = do
  putStrLn "put a number, -1 to terminate. "
  age <- checkNumber <$> getLine
  if age == -1
  then return []
  else do 
          list <- getAge
          return (age : list)

checkNumber :: String -> Int
checkNumber n = Safe.readNote "Not a number" n
-- or Safe.readDef 0 n

getAverage xs = sum xs `div` length xs

-- IO interactive program template from Erik's online class

--interactive prelude evaluate terminate success next word = do
--  prelude
--  xs <- getLine
--  if evaluate xs word
--  then terminate xs
--  else do success word xs
--          interactive prelude evaluate terminate success next (next word)
--  

  
  
